core/src/apps/zcash/signer.py

```diff
@@ -0,0 +1,155 @@
+from . import sig_hasher
+
+from micropython import const
+
+from trezor import log
+from trezor.utils import ensure
+from trezor.wire import ProcessError, DataError
+
+from apps.bitcoin.sign_tx.bitcoinlike import Bitcoinlike
+from apps.bitcoin.sign_tx import approvers
+from apps.common.writers import (
+    write_compact_size,
+    write_uint32_le,
+)
+
+from typing import TYPE_CHECKING
+if TYPE_CHECKING:
+    from typing import Sequence, Tuple
+    from apps.common.coininfo import CoinInfo
+    from apps.bitcoin.sign_tx.tx_info import OriginalTxInfo, TxInfo
+    from apps.bitcoin.writers import Writer
+    from apps.bitcoin.sign_tx.sig_hasher import SigHasher
+    from apps.bitcoin.sign_tx.approvers import Approver
+    from trezor.utils import HashWriter
+    from trezor.messages import (
+        PrevTx, SignTx, TxInput, TxOutput,
+        TxAckInput, TxInfo, OriginalTxInfo,
+    )
+
+OVERWINTERED = const(0x8000_0000)
+
+class Zcash(Bitcoinlike):
+    def __init__(
+        self,
+        tx: SignTx,
+        keychain: Keychain,
+        coin: CoinInfo,
+        approver: Approver,
+    ) -> None:
+        ensure(coin.overwintered)
+        if tx.version != 5:
+            # seemless version upgrade
+            log.warning(__name__, "Transaction format version {} is not supported.".format(tx.version))
+            log.warning(__name__, "Switching to version 5.".format(tx.version))
+            tx.version = 5
```

Comment:
After discussion with @matejcik, we don't think that changing the version here is a good idea. If the version is not 5, then the signing should fail, because it indicates that the host application is expecting a different transaction version than what Trezor is going to be providing. BTW, note that the warning would only get logged in the debug firmware.

Reply:
https://github.com/trezor/trezor-firmware/pull/2166/commits/4b94cad001411de4a84864be9fa678c53f74025d